SshAgentLib/UnixSocket.cs

```diff
@@ -0,0 +1,221 @@
+    List<Socket> clientSockets = new List<Socket>();
+    object clientSocketsLock = new object();
+
+    public delegate void ConnectionHandlerFunc(Stream stream, Process process);
+    public ConnectionHandlerFunc ConnectionHandler { get; set; }
+
+    /// <summary>
+    /// Create new "unix domain" socket for use with Linux
+    /// </summary>
+    /// <param name="path">The name of the file to use for the socket</param>
+    public UnixSocket(string path)
+    {
+      this.path = path;
+      guid = Guid.NewGuid();
+      {
+        try {
+          socket = new Socket(AddressFamily.Unix, SocketType.Stream,
+            ProtocolType.Unspecified);
+          var endpoint = new UnixEndPoint(path);
+          sockaddr = endpoint.Serialize();
+          socket.Bind(endpoint);
+          var fileSecurity = File.GetAccessControl(path);
+          // This turns off ACL inheritance and removes all inherited rules
+          fileSecurity.SetAccessRuleProtection(true, false);
+          // We are left with no permissions at all, so we have to add them
+          // back for the current user
+          var userOnlyRule = new FileSystemAccessRule(
+            WindowsIdentity.GetCurrent().User,
+            FileSystemRights.FullControl,
+            AccessControlType.Allow);
+          fileSecurity.SetAccessRule(userOnlyRule);
+          File.SetAccessControl(path, fileSecurity);
+          socket.Listen(5);
+          var socketThread = new Thread(AcceptConnections);
+          socketThread.Name = "UnixSocket";
+          socketThread.Start();
+        } catch (Exception) {
+          if (socket != null)
+            socket.Close();
+          File.Delete(path);
+          throw;
+        }
+      }
+    }
+
+    /// <summary>
+    /// Tests a file to see if it looks like a Unix socket file
+    /// </summary>
+    /// <param name="path">The path to the file.</param>
+    /// <returns><c>true</c> if the file contents look correct</returns>
+    public static bool TestFile(string path)
+    {
+      var info = new FileInfo(path);
+      return info.Length == 0;
+    }
+
+    public void Dispose()
+    {
+      Dispose(true);
+      GC.SuppressFinalize(this);
+    }
+
+    void Dispose(bool disposing)
+    {
+      if (!disposed) {
+        disposed = true;
+        if (disposing) {
+          // Dispose managed resources
+          foreach (var clientSocket in clientSockets) {
+            clientSocket.Dispose();
+          }
+          socket.Dispose();
+          File.Delete(path);
+        }
+        // Dispose unmanaged resources
+      }
+    }
+
+    void AcceptConnections()
+    {
+      var buffer = new byte[16];
+      while (true) {
+        try {
+          BindingFlags
+            sbinding = BindingFlags.Static | BindingFlags.NonPublic | BindingFlags.Public,
+            ibinding = BindingFlags.Instance | BindingFlags.NonPublic | BindingFlags.Public;
+          var m_Size = typeof(SocketAddress).GetField("m_Size", ibinding);
```

Comment:
Why are we using reflection here?

Reply:
IIRC there's no way to set that field publicly.

Reply:
Can we add a comment explaining why we are doing this and why it is safe to do this?

Reply:
Sure.

Reply:
I'm looking over this again and I'm honestly not remembering why I had to set `m_Size` at all. I vaguely seem to recall seeing something funny happening when a method was called (it might that field might have been reset when `Accept` was called?) and that it took me a long time to track it down, but when I remove the call to `SetValue` now, I don't appear to see the issue. It's possible it I may have seen this issue on an older system, but I'm not sure. Do you think I should keep it (and leave a comment) or just take it out?

Reply:
Oh I just remembered `Accept()` is supposed to return the address as well, so that needs to be used when creating the `EndPoint`. Probably better to keep it I think; I'll add a comment.